app/Platform/Enums/HashCompatibility.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Platform\Enums;
+
+abstract class HashCompatibility
+{
+    public const Unverified = 'unverified';
+
+    public const InProgress = 'in_progress';
```

Comment:
What does "in_progress" mean? Is compatibility being added or tested? Should we have separate values for both of those states? 

Wouldn't a hash be unverified until it was tested, at which point it transitions to verified or problematic?

Reply:
I brought these over from https://discord.com/channels/310192285306454017/1085433484048343050

To quote @televandalist:

>🟡 Unverified: The hash has not been verified by an admin (or "entry manager" whenever site perms get an overhaul).
🔵 In Progress: The hash is in the current batch of hashes being verified.
🟢 Verified: An admin/entry manager has confirmed the hash is legitimate and info/metadata is accurate.
🔴 Problematic (or whatever term works best here): This would basically be like opening a ticket on a hash. The most common cause of this would be if an admin/entry manager is unable to reproduce the hash. A bad base could have been used, there could have been an issue creating the patch, etc.

Reply:
Oh, that's not compatibility, that's actually validity.

I was thinking:
* Unverified = Untested
* In Progress = Being tested (or being made compatible)
* Verified = Testing complete, version associated to hash is compatible
* Problematic = Testing complete, version is not compatible and should not be used to earn achievements.

We may want to track both, but I think we can combine them:
* Unverified = This hash hasn't been looked at
* Unknown = We couldn't figure out where this hash came from
* Untested = Known hash, but compatibility hasn't been tested (or set doesn't exist)
* Compatible = Known hash, works with the set
* Incompatible = Known hash, does not work with the set

I think "In Progress" is unnecessary - even if multiple people are trying to identify hashes independently, they'll probably do so one at a time, so the user could change it over to Unknown while they are doing the identification.

Reply:
When I'm working on hash/patch stuff, this is my usual process:
- Download an updated .csv of [this query](https://metabase.retroachievements.org/question/234-hash-database) and import into my hash sheet.
- Compare hashes in my hash sheet to the new query to see if anything has been removed.
- Compare hashes in the new query to see what I already have logged and delete those rows. Leftovers are hashes that have been linked since my last update.
- Compare new hashes to ones I have listed in the [RA Hashes repository](https://github.com/RetroAchievements/RAHashes). I start with No Intro and Redump, then work my way down the line. If file names and labels check out, then I'll move those to the "Verified" tab in my sheet.
- Leftovers from that are moved to a "ToDo" tab. These are what I consider "Unverified" and are almost always homebrews, unlicensed, pirate, and ROMs that require a patch of some sort.
  - For those that require patches, I either obtain or make the required patches and put them in the RAPatches repo. Once I reproduce a linked hash, I adjust the file name and label(s) as needed. I then update that hash in my sheet and move it to an "RAPatches" tab, which I also consider "Verified"
  - For select homebrews and other "aftermarket" games, I periodically submit information to No Intro so they end up in their DATs. I've done this for several hundred hashes at this point 😅

So, *to me:*
- **Verified** = A hash's filename and label(s) match those of a DAT by No Intro, Redump, Final Burn Neo, TOSEC, etc. If it requires a patch, then the patch is in the RAPatches repo and I have confirmed the hash is reproduceable. Over 34,000 hashes in our DB match this criteria.
- **Unverified** = The hash does not match one of the above and I have not tried to reproduce. There are currently a little over 200 of these. A small handful of these are just subset patches that haven't been submitted yet.
- **In Progress** = The hash does not match one of the above. I was able to reproduce, but have not added a patch to the RAPatches repo or submitted to No Intro's DATs yet. A lot of these are just GoodTools and NonGood hashes that need to be converted to No Intro + RAPatches. There are currently around 900 of these.
- **Problematic** = I was not able to reproduce the hash. This is very rare and I think there are only a small handful at any given time. These are typically versions of newer hacks or homebrews that I didn't obtain a patch for before an update or just something completely batshit.

Anyway, in the overall scope of things, I agree with Jamiras that "In Progress" is probably unnecessary. I have a tendency to think about this stuff in regards to my own workflow since I've been doing 99% of it for a few years now. 

As for "Compatible" and "Incompatible," I'm not sure. If something is incompatible it shouldn't be linked at all. Or are we planning to store hash info even for hashes that aren't linked? I'd love that, personally.

Reply:
I'd like to be able to link incompatible hashes, which could be leveraged to return a pseudo-set to the client that would inform the user that they're using a known, but incompatible ROM instead of just saying "no achievements found" or "unknown game" like it does currently.

See #717 and #789.

Reply:
Would we go about that by making dummy entries like the "Cheated ROM" ones? Elmo and I discussed doing that a few years ago, but I think we were waiting until the hash database to be in a better spot. I'd say it's manageable enough now. Could even rename the cheat entries and continue using those. I think people get freaked out when loading them as is 😆 

I could run our hash list against a P/C DAT and knock out a ton of them pretty quick.

Though my usual method of mass linking hashes (typically for rollouts and to create hub entries) leads to a 403 error now....

Reply:
I'm thinking more along the lines of:

| Hash | Compatibility | Label |
| ---- | ---- | ---- |
| abcdef1234567890 | Compatible | Game (E).rom |
| 1a2b3c4d5e6f7890 | Incompatible | Game (J).rom |
| 88347abcdea9128a | Untested | Game (U).rom |
| 2abbabbab2babab2 | Incompatible | Game (E) [Infinite lives cheat].rom |

We might even want to go so far as to have a Blocked compatibility that would behave like Incompatible, but not be shown on the linked hashes page.

Reply:
Updated in latest:

```php
public const Unverified = 'unverified';
public const Unknown = 'unknown';
public const Untested = 'untested';
public const Compatible = 'compatible';
public const Incompatible = 'incompatible';
```

Reply:
Does there need to be a query/command to initialize all of the existing records to 'unverified'? Or will null naturally map to that later? We can probably address that in whatever PR adds the ability to edit these.

Reply:
Indeed, I plan on taking care of working with this value in a subsequent PR.